### PR TITLE
RFC-0001 Polling mechanism for the branch-based planner

### LINKED ISSUE #595: Write an RFC describing the security considerations for branch-based planner
Relates to #527 

The RFC should:
- Evaluate alternative approaches for communicating with git providers (i.e webhook vs polling) and their security posture
- Propose an approach and justify it
- Propose an architecture that supports the chosen approach

### PR DESCRIPTION
**Pull request by chanwit**
This RFC proposes a polling mechanism as a secure and portable solution for tracking changes in the
Branch-based Planner with GitHub. The proposed system reduces security risks associated with webhooks, which require
exposing a publicly accessible endpoint by having the Kubernetes cluster periodically query GitHub for changes.

Fixes #595 

Per the RFC process, the proposal can only be merged after it has been approved by either
-  two project maintainers, or
-  one project maintainer and one individual from the following: a Weaveworks Tech Lead, a Weaveworks Product Manager, or a Weaveworks Principal Engineer.

### CODE REVIEW

**Comment by yitsushi** [comment]
For polling, do we need an extra CRD that describes a repo to watch? Something like this:

```yaml
---
apiVersion: infra.contrib.fluxcd.io/v1alpha2
kind: TerraformBranchBasedPlanner
metadata:
  name: helloworld-tf
  namespace: flux-system
spec:
  interval: 1m
  sourceRef:
    kind: GitRepository
    name: helloworld
    namespace: flux-system
```

File: rfcs/0001-polling-mechanism/README.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+On the contrary, the polling mechanism works largely the same way for any Git provider. We need to make requests to the
+provider's API to fetch the required data, which is typically available through similar RESTful APIs across different
+providers. Therefore, we can easily adapt our polling code to a new provider by changing the API endpoints and adjusting
+for any differences in the API responses.
+
+In addition, the polling mechanism allows us to control the rate of requests, which can be beneficial for dealing with
+rate limits imposed by different providers. With webhooks, the rate of incoming messages is determined by the activity
+on the Git provider, which could potentially overwhelm our system or cause us to exceed rate limits.
+
+## Proposed solution
```
Comment (yitsushi) [inline rfcs/0001-polling-mechanism/README.md:82]: For polling, do we need an extra CRD that describes a repo to watch? Something like this:

```yaml
---
apiVersion: infra.contrib.fluxcd.io/v1alpha2
kind: TerraformBranchBasedPlanner
metadata:
  name: helloworld-tf
  namespace: flux-system
spec:
  interval: 1m
  sourceRef:
    kind: GitRepository
    name: helloworld
    namespace: flux-system
```

Comment (chanwit) [reply rfcs/0001-polling-mechanism/README.md:82]: I would go with the simplest implementation first without introducing a new CRD.

Comment (yitsushi) [reply rfcs/0001-polling-mechanism/README.md:82]: Hmm... and what would be the "simplest implementation"?

Comment (chanwit) [reply rfcs/0001-polling-mechanism/README.md:82]: A simple polling, PR query, etc without CRD - that's what I meant.

Comment (yitsushi) [reply rfcs/0001-polling-mechanism/README.md:82]: Yes, but how the system will know what repositories it has to query for PRs?

Comment (chanwit) [reply rfcs/0001-polling-mechanism/README.md:82]: Would you please create a spike ticket and move this discussion there? It's a bit out of scope for this RFC.

Comment (yitsushi) [reply rfcs/0001-polling-mechanism/README.md:82]: Yes, but I don't think it's fully out of scope as it's probably a "drawback" as we have to either create a CRD or iterate over all Terraform resources on the cluster, parse their repo, aggregate the repositories (so no duplicates) and then we can query PRs. Compared to webhook where we know the repository from received payload.

Comment (chanwit) [reply rfcs/0001-polling-mechanism/README.md:82]: > Compared to webhook where we know the repository from received payload.

I mistook your concern then. 

We have that information in the code provided. See example 1.

Comment (yitsushi) [reply rfcs/0001-polling-mechanism/README.md:82]: In the provided code we have a hard-coded repository.
```go
prs, _, _ := client.PullRequests.List(ctx, "weaveworks", "tf-controller", nil)
```

Comment (yitsushi) [reply rfcs/0001-polling-mechanism/README.md:82]: ConfigMap sounds good.

Comment (chanwit) [reply rfcs/0001-polling-mechanism/README.md:82]: Scoping down to a set of specified repositories would also improve security.

Thank you for bringing this up @yitsushi :+1:

**Review by yitsushi — APPROVED** [review]
lgtm

**Review by yiannistri — APPROVED** [review]

**Pull request merged**